section_synonym.go

```diff
@@ -0,0 +1,740 @@
+//  Copyright (c) 2024 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zap
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+
+	"github.com/RoaringBitmap/roaring"
+	"github.com/RoaringBitmap/roaring/roaring64"
+	index "github.com/blevesearch/bleve_index_api"
+	seg "github.com/blevesearch/scorch_segment_api/v2"
+	"github.com/blevesearch/vellum"
+)
+
+func init() {
+	registerSegmentSection(SectionSynonymIndex, &synonymIndexSection{})
+	invertedIndexExclusionChecks = append(invertedIndexExclusionChecks, func(field index.Field) bool {
+		_, ok := field.(index.SynonymField)
+		return ok
+	})
+}
+
+// -----------------------------------------------------------------------------
+
+type synonymIndexOpaque struct {
+	results []index.Document
+
+	// indicates whether the following structs are initialized
+	init bool
+
+	// FieldsMap maps field name to field id and must be set in
+	// the index opaque using the key "fieldsMap"
+	// used for ensuring accurate mapping between fieldID and
+	// thesaurusID
+	//  name -> field id
+	FieldsMap map[string]uint16
+
+	// ThesaurusMap adds 1 to thesaurus id to avoid zero value issues
+	//  name -> thesaurus id + 1
+	ThesaurusMap map[string]uint16
+
+	// ThesaurusMapInv is the inverse of ThesaurusMap
+	//  thesaurus id + 1 -> name
+	ThesaurusInv []string
+
+	// Thesaurus for each thesaurus ID
+	//  thesaurus id -> LHS term -> synonym postings list id + 1
+	Thesauri []map[string]uint64
+
+	// LHS Terms for each thesaurus ID, where terms are sorted ascending
+	//  thesaurus id -> []term
+	ThesaurusKeys [][]string
+
+	// FieldIDtoThesaurusID maps the field id to the thesaurus id
+	//  field id -> thesaurus id
+	FieldIDtoThesaurusID map[uint16]int
+
+	// SynonymIDtoTerm maps synonym id to term for each thesaurus
+	//  thesaurus id -> synonym id -> term
+	SynonymTermToID []map[string]uint32
+
+	// SynonymTermToID maps term to synonym id for each thesaurus
+	//  thesaurus id -> term -> synonym id
+	// this is the inverse of SynonymIDtoTerm for each thesaurus
+	SynonymIDtoTerm []map[uint32]string
+
+	//  synonym postings list -> synonym bitmap
+	Synonyms []*roaring64.Bitmap
+
+	builder    *vellum.Builder
+	builderBuf bytes.Buffer
+
+	tmp0 []byte
+
+	thesaurusAddrs map[int]int
+}
+
+func (so *synonymIndexOpaque) Set(key string, value interface{}) {
```

Comment:
The exported APIs at the very least require code commentary, some code commentary around the others is a plus too.

Reply:
ok